### PR TITLE
feat(bench): the tuning loop — demo tasks as coding projects, cost and parallelism per task, deltas per cycle

### PR DESCRIPTION
**Pull request by yalexx**
## What

The bench's **tuning loop** (`bench/loop.mjs`), the owner's ask: a main loop that starts the bench, triggers the demo tasks as coding PROJECTS, and collects what the harness is tuned by — token spend, parallelisation, time to finish and cost per task — cycle after cycle.

- Runs the suite's tasks as `bench-<task>-<stamp>` folders directly under the owner's project folder, so each is a project the Coding Agent app lists (the runner kept a folder of its own and switched the box's project folder to it, which hid the owner's projects meanwhile; the loop switches only with `--workroot`, and says so).
- Samples every run every 5 s while it works (`samples.json` in the run's results directory) → **parallelism**: peak helpers at once, helper-seconds, the share of the clock with a helper out, and agent-seconds per wall-second (1.0 = the main loop alone).
- **Cost per task** from `bench/pricing.json` (USD per million tokens by model; the shipped numbers are PLACEHOLDERS to set to the plan this box is on) over the per-model usage the capture already sums from the transcript; a model the table lacks is flagged as unpriced, never counted as free.
- One report per cycle (`results/<suite>/report-<label>.md`): the figures per task, the change against the previous cycle (`--baseline`, or the cycle before within `--cycles N`), and a "look at" list of plain rules (a run that did not complete, a low score, refused actions, a long run with no helper, a burst of helpers that sat idle, a thinking share over 30%). `loop-<label>.jsonl` carries the figures per run.
- The pure parts (`lib/cost.mjs`, `lib/metrics.mjs`) are unit-tested (`src/tests/unit/bench-metrics.test.ts`); the README gains a Loop section.

Smoke on the box (2026-09-05): `node bench/loop.mjs --tasks s-01-single-edit --label smoke` → the task ran as project `bench-s-01-single-edit-…`, completed in 15 s, score 100, 48k tokens, $0.021 at the placeholder prices, report written.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013Fz3uZVzYQzcNAkw1cQ8ff

## Summary by CodeRabbit

* **New Features**
  * Added a command-line benchmark workflow supporting task runs, repeated cycles, baselines, dry runs, workspace selection, progress sampling, scoring, and reports.
  * Added benchmark metrics for duration, token usage, parallelism, cycle comparisons, tuning hints, and task-level results.
  * Added pricing configuration, cost reporting, and persistent benchmark result tracking, including handling for missing or invalid pricing data and failed runs.

* **Documentation**
  * Documented the benchmark workflow, reports, pricing behavior, and workspace layout.

* **Tests**
  * Added coverage for benchmark metrics, pricing calculations, formatting, persistence, defaults, and edge cases.

### CODE REVIEW

File: bench/lib/cost.mjs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    if (!rates) {
+      out.unpriced.push(model);
+      out.byModel[model] = { usd: null, priced: false, tokens: input + output + cacheRead + cacheWrite };
+      continue;
+    }
+    const usd = (input * rates.input + output * rates.output + cacheRead * rates.cacheRead + cacheWrite * rates.cacheWrite) / PER_MILLION;
+    out.byModel[model] = { usd: round6(usd), priced: true, tokens: input + output + cacheRead + cacheWrite };
+    out.totalUsd += usd;
+  }
+  out.totalUsd = round6(out.totalUsd);
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/lib/cost.mjs]: Done: `costOfUsage` answers `pricedUsd` and a `totalUsd` that is null while any model is unpriced; the summary, the report and the console say so — test updated.

File: bench/lib/metrics.mjs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      : null,
+  };
+}
+
+/**
+ * What changed per task between two cycles — the loop's reading. Positive
+ * `pct` is up; for wall, tokens and cost that is worse, for score better.
+ */
+export function deltaByTask(previous, current) {
+  const prev = new Map((previous ?? []).map((r) => [r.task, r]));
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/lib/metrics.mjs]: Done: figures carry `rep`, `deltaByTask` keys on `task#rep` (`figureKey`) — tested with a repeated task.

File: bench/loop.mjs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+ * The task's folder IS the project: `<projects>/bench-<task>-<stamp>`, directly
+ * under the owner's project folder, which is what the app lists. A task's
+ * "outside" files (the refusal probe's `../shared-config/…`) land beside it.
+ */
+function seedProject(task, projectsRoot, stamp) {
+  const workdir = path.join(projectsRoot, `bench-${task.id}-${stamp}`);
+  fs.mkdirSync(workdir, { recursive: true });
+  if (task.seedDir) fs.cpSync(task.seedDir, workdir, { recursive: true });
+  for (const [rel, content] of Object.entries(task.outside ?? {})) {
+    const dest = path.resolve(workdir, "..", rel);
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs]: Done: outside files resolve from the project folder (`path.resolve(workdir, rel)`).

File: bench/loop.mjs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+  const file = path.join(RESULTS_DIR, suiteVersion, `report-${label}.md`);
+  fs.writeFileSync(file, lines.join("\n") + "\n");
+  return file;
+}
+
+async function runCycle({ args, suite, tasks, label, projectsRoot, pricing }) {
+  const figures = [];
+  const plan = tasks.flatMap((t) => Array.from({ length: args.repeat }, (_, i) => ({ task: t, rep: i + 1 })));
+  for (const { task, rep } of plan) {
+    const stamp = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs]: Done: workspace stamps carry four random hex characters, cycle labels carry the time to the second.

File: bench/loop.mjs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        run = await sampleUntilSettled(runId, Date.now() + 30_000, samples);
+      }
+      settled = await waitForCommit(runId);
+      if (settled.run) run = settled.run;
+    } catch (err) {
+      outcome = "transport-failure";
+      console.error(`  polling failed after start: ${String(err).slice(0, 200)}`);
+    }
+    const score = await scoreTask(task, workdir, run);
+    const wallMs = (run.completedAt ?? Date.now()) - run.startedAt;
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs:209]: Done: `commitLagMs` is persisted and `endToEndMs` (settle + lag) is the time the report leads with, with the settle time beside it — tested.

File: bench/loop.mjs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+function readFigures(suiteVersion, label) {
+  const file = path.join(RESULTS_DIR, suiteVersion, `loop-${label}.jsonl`);
+  if (!fs.existsSync(file)) return null;
+  return fs.readFileSync(file, "utf8").split("\n").filter(Boolean).map((l) => JSON.parse(l));
+}
+
+/** A stamp no two runs share: the second, and four random hex characters after it. */
+function stamp() {
+  return `${new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19)}-${crypto.randomBytes(2).toString("hex")}`;
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs]: Done: `stamp()` carries 16 random bytes and the cycle label is built from it.

File: bench/loop.mjs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+async function runCycle({ args, suite, tasks, label, projectsRoot, pricing }) {
+  const figures = [];
+  const plan = tasks.flatMap((t) => Array.from({ length: args.repeat }, (_, i) => ({ task: t, rep: i + 1 })));
+  for (const { task, rep } of plan) {
+    const workdir = seedProject(task, projectsRoot, stamp());
+    console.log(`\n▶ ${task.id} rep ${rep} as project ${path.basename(workdir)}`);
+    const started = await startRun({ task: task.brief, directory: workdir });
+    if (started.status !== 202 || !started.json?.run) {
+      console.error(`  run refused: ${started.status} ${started.text.slice(0, 300)}`);
+      figures.push(taskFigures({ line: { task: task.id, tier: task.tier, runId: null, outcome: "not-started", wallMs: null, subagentsByType: {}, modelsUsed: [] }, cost: null, parallel: parallelism([], 0), cycle: label, rep }));
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs]: Done: a busy refusal is recorded with `costOfUsage(null, pricing)` — zero usage, zero cost — and a test pins a priced run beside a not-started one summing to a number.

File: bench/loop.mjs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+  const plan = tasks.flatMap((t) => Array.from({ length: args.repeat }, (_, i) => ({ task: t, rep: i + 1 })));
+  for (const { task, rep } of plan) {
+    const workdir = seedProject(task, projectsRoot, stamp());
+    console.log(`\n▶ ${task.id} rep ${rep} as project ${path.basename(workdir)}`);
+    const started = await startRun({ task: task.brief, directory: workdir });
+    if (started.status !== 202 || !started.json?.run) {
+      console.error(`  run refused: ${started.status} ${started.text.slice(0, 300)}`);
+      // A run that never started spent nothing: zero usage priced as zero,
+      // so the cycle's cost stays a number rather than "n/a".
+      figures.push(taskFigures({ line: { task: task.id, tier: task.tier, runId: null, outcome: "not-started", wallMs: null, subagentsByType: {}, modelsUsed: [] }, cost: costOfUsage(null, pricing), parallel: parallelism([], 0), cycle: label, rep }));
```
[the remark this answers is not in the document]
Comment (yalexx) [reply bench/loop.mjs]: Done: one `record()` appends every figure — a refused run included — through `appendFigure` (bench/lib/figures-file.mjs), and a test round-trips a started + a not-started figure through the file and the summary.
